README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Purpose of corr2D
=================

[![CRAN version](http://www.r-pkg.org/badges/version/corr2D)](https://cran.r-project.org/package=corr2D) ![cranlogs](http://cranlogs.r-pkg.org./badges/corr2D) [![Build Status](https://travis-ci.com/clacor/corr2D.svg?branch=master)](https://travis-ci.com/clacor/corr2D)

The package corr2D implements two-dimensional (2D) correlation analysis in R. It allows the user to calculate and plot homo as well as hetero 2D correlation spectra from an input matrix containing the (preprocessed) spectra. The user can freely choose the reference spectrum used to create the dynamic spectra, how the perturbation variables vector should be handled, if and how the resulting 2D spectra should be scaled and how many cores should be used to calculate the correlation matrix.
The plot functions offer a wide variety of options to customize the look of 2D spectra. In this context the appearance of the main spectrum can be altered and the 1D reference spectra as well as the color legend can be controlled.
The package also features an example dataset containing preprocessed FT-Raman spectra of a self-healing polymer sample.

How to use corr2D on a simple example
=====================================

First we need to make the package available and get an example dataset. For illustration purposes we use the FuranMale dataset from the present corr2D package.

``` r
library("corr2D")
data(FuranMale, package = "corr2D")
```

As the next step we calculate the homo 2D correlation spectra of the FuranMale dataset using the function corr2d() and the first spectrum of the dataset as reference spectrum.

``` r
twod <- corr2d(FuranMale, Ref1 = FuranMale[1, ], corenumber = 1)
#> HOMO-Correlation: 1 cores used for calculation
#> 11:04:04 - Fast Fourier Transformation and multiplication 
#>  to obtain a 145 x 145 correlation matrix 
#> 11:04:04 - Done
```

The last step is to plot the resulting 2D correlation spectra. As an example we plot the synchronous correlation spectrum of the FuranMale dataset. The appearance of the 2D plot can be altered by i.e. using the xlab and ylab options to control the labels at the x- and y-axes.

``` r
plot_corr2d(twod, xlab = expression(paste("relative Wavenumber" / cm^-1)),
                  ylab = expression(paste("relative Wavenumber" / cm^-1)))
#> [1] 3
```

![](README-unnamed-chunk-4-1.png)

For an in-depth tutorial as well as a detailed description of the code see R. Geitner et al. (2019) <DOI: https://doi.org/10.18637/jss.v090.i03>.


NEWS.md
# corr2D 1.0.3
- Moved package profr and xtable to Suggests. They are optionally used to check
    corr2ds performance and are thus not essential to the
    package.

# corr2D 1.0.1
- Moved package rgl to Suggests as the future of OpenGL is uncertain. rgl is
    used as an optional preview in corr2d and is thus not essential to the
    package.

# corr2D 1.0.0
- Added publication in Journal of Statistical Software as preferred
    citation. The publication is a good start for newcomers to the field of
    2D correlation spectroscopy (in R)!

# corr2D 0.4.0
- Added internal 'testthat' tests to corr2D to easier check correlation results
    during package development.
- Added a default print() method for corr2d objects (plot.default() for now).
- Changed plot_corr2d() to use plot() instead of plot.default() to enable
    method dispatch.
- Fixed a small error with the calculation of reference spectrum 2 (Ref2) if
    reference spectrum 1 (Ref1) is not present (Thanks to Bettina Gruen).

# corr2D 0.3.0
- Added 2T2D correlation analysis to the package. Function uses the approach
    described by I. Noda (2018)
    <DOI: https://doi.org/10.1016/j.molstruc.2018.01.091>.
- A freshly published paper at The Journal of Statistical Software was added
    as a vignette. This will be updated in the future to also include
    descriptions of new functions.
- Added citation referencing the published paper at The Journal of Statistical
    Software.

# corr2D 0.2.0
- Added 2D codistribution analysis to the packages functions. Function uses
    the approach described by I. Noda (2014)
    <DOI: https://doi.org/10.1016/j.molstruc.2014.01.024>.
- Rewrote the function plot_corr2d() for better control about the plot
    appearance. Introduced the graphical parameters "lwd", "lwd.axis",
    "lwd.spec", "col", "col.axis", "col.lab", "cex.axis", "cex.lab", "cex.leg",
    "font.axis" and "font.lab" which are/are derived from "par".
    "at.xaxs"/"at.yaxs" and "label.xaxs"/"label.yaxs" allow control over the
    axes ticks and their labels. "line.xlab"/"line.ylab" control the position
    of the axes label.

# corr2D 0.1.12
- Rewrote the function plot_corr2d(): The graphical parameters specified at
    "..." are now partly transferred to all parts of the plot, not just the
    main screen.
- Cleaned up the image.plot() code inside plot_corr2d().
- Fixed a bug which prevented the use of an individual color palette in
    plot_corr2din3d().
- Changed the default color palette in plot_corr2din3d() from
    fields::timcolors() to colorspace::diverge_hcl(). This change should help
    to improve the default graphics quality.

# corr2D 0.1.11
- Corrected the default value for the normalization factor. It's now the number
    of perturbation variables which is NROW() of Mat1.
- Added error massages to all functions.
- Rewrote the function sim2ddata() to allow to set C and Camp NULL. In that
    case only the first order reaction A -> B will be simulated and sampled.

# corr2D 0.1.10
- Added the function sim2ddata to simulate artificial data.
- Fixed the plotting of the reference spectra on the 2D correlation spectra.
    Before the reference spectra used an arbitrary x-axis. Now they use their
    correct spectral variables for plotting. 2D correlation peaks should now
    align with their 1D counterparts in the reference spectra.
- Rewrote the interpolation of the perturbation variables to no longer use the
    ineffective cbind in a for loop. Instead apply and sapply are now used
    instead.
- The interpolated perturbation variables are now saved when the perturbation
    values get interpolated.
- Cleaned up some left/right confusions in the documentation of plot_corr2d.

# corr2D 0.1.9
- Package is released on CRAN

## Known Bugs
- A new plot after a call of plot_corr2d() is plotted inside the main part.
    It's a consequence of keeping the split.screen of the main part active
    for interactive data readout and manipulation.

tests/testthat.R
library(testthat)
library(corr2D)

test_check("corr2D")


tests/testthat/test_corr2d_numeric.R
context("corr2d numeric correctness")
library("corr2D")

test_that("corr2d gives correct numeric result", {
  expect_equal(corr2d(matrix(c(1:3, 3:1), 3, 2), Wave1 = 1:2, Ref1 = rep(0, 2), corenumber = 1)$FT[1, 2], 30/(2*pi) + 0i)
  expect_equal(corr2d(matrix(c(1:3, 1:3), 3, 2), Wave1 = 1:2, corenumber = 1)$FT[1, 2], 6/(2*pi) + 0i)
  expect_equal(corr2d(matrix(c(1:3, 1:3), 3, 2) * 10, Wave1 = 1:2, corenumber = 1)$FT[1, 2], corr2d(matrix(c(1:3, 1:3), 3, 2), Wave1 = 1:2, corenumber = 1)$FT[1, 2] * 100)
})

R/codys2d.R
#' Two-dimensional codistribution spectroscopy.
#'
#' \code{codis2d} calculates the synchronous and asynchronous codistribution
#'     spectra.
#'
#' \code{codis2d} calculates the the synchronous 2D correlation spectrum and
#'     uses the 2D spectrum to calculate the synchronous and asynchronous
#'     codistribution spectra. For parallelization the
#'     \code{\link[parallel]{parCapply}} function is used. Large input matrices
#'     (> 4000 columns) can lead to long calculation times depending on the
#'     number of cores used. Also note that the resulting matrix can become
#'     very large, adjust the RAM limit with \code{\link[utils]{memory.limit}}
#'     accordingly. For a detailed description of the underlying math see
#'     references.
#'
#' @param Mat Numeric matrix containing the data which will be correlated;
#'     '\emph{spectral variable}' by columns and '\emph{perturbation variables}'
#'     by rows.
#' @param Ref Numeric vector containing a single spectrum, which will be
#'     subtracted from \code{Mat} to generate dynamic spectra for 2D
#'     correlation analysis. Default is \code{NULL} in which case the
#'     \code{colMeans()} of \code{Mat} is used as reference. The length of
#'     \code{Ref} needs to be equal to the number of columns in \code{Mat}.
#'     2D codistribution spectroscopy is only strictly defined using the
#'     perturbation-mean spectrum as reference spectrum. Thus, any deviation
#'     from this definition can lead to unexpected results.
#' @param Wave Numeric vector containing the spectral variable. Needs to be
#'     specified if column names of \code{Mat} are undefined.
#' @param Time Numeric vector containing the perturbation variables. If
#'     specified, \code{Mat} will be interpolated to \code{N} equally spaced
#'     perturbation variables using \code{Int}.
#' @param Int Function specifying how the dataset will be interpolated to give
#'     \code{N} equally spaced perturbation variables. \code{\link[stats]{splinefun}}
#'     (default) or \code{\link[stats]{approxfun}} can for example be used.
#' @param N Positive, non-zero integer specifying how many equally spaced
#'     perturbation variables should be interpolated using \code{Int}. \code{N}
#'     should be higher than 4. \code{corr2d} is fastest if \code{N} is a power
#'     of 2.
#' @param Norm A number specifying how the correlation matrix should be
#'     normalized.
#' @param scaling Positive real number used as exponent when scaling the dataset
#'     with its standard deviation. Defaults to 0 meaning no scaling. 0.5
#'     (\emph{Pareto scaling}) and 1 (\emph{Pearson scaling}) are commonly used
#'     to enhance weak correlations relative to strong correlations. 2D
#'     codistribution spectroscopy is only strictly defined without the usage
#'     of any scaling techniques. Thus, any deviation from this definition can
#'     lead to unexpected results.
#' @param corenumber Positive, non-zero integer specifying how many CPU cores
#'     should be used for parallel fft computation.
#' @param preview Logical: Should a 3D preview of the asynchronous codistribution
#'     spectrum be drawn at the end? Uses \code{\link[rgl]{persp3d}} from \pkg{rgl}
#'     package.
#'     
#' @return \code{codis2D} returns a list of class "corr2d" containing the complex
#'     codistribution matrix (\code{$FT}), the synchronous correlation spectrum
#'     ($corr), the used reference spectrum \code{$Ref1} and \code{$Ref2}, the 
#'     spectral variables \code{$Wave1} and \code{$Wave2} as well as the
#'     (interpolated) perturbation variables (\code{$Time}).
#' 
#' @references 
#'     I. Noda (2014) <DOI:10.1016/j.molstruc.2014.01.024>
#'     
#' @seealso For plotting of the resulting list containing the 2D codistribution
#'     spectra see \code{\link{plot_corr2d}} and \code{\link{plot_corr2din3d}}.
#' 
#' @examples
#'     testdata <- sim2ddata(C = NULL, Camp = NULL)
#'     codis <- codis2d(testdata, corenumber = 1)
#'     
#'     plot_corr2d(codis, Im(codis$FT),
#'                 xlab = expression(paste("Wavenumber" / cm^-1)),
#'                 ylab = expression(paste("Wavenumber" / cm^-1)))
#'
#' @aliases codis2d
#' 
#' @export
#' @importFrom stats sd
codis2d <-
    function(Mat, Ref = NULL, Wave = NULL, Time = NULL, Int = stats::splinefun,
             N = 2^ceiling(log2(NROW(Mat))), Norm = 1/(NROW(Mat) - 1), 
             scaling = 0, corenumber = parallel::detectCores(), preview = FALSE)
    {
        # check user input for errors -----------------------------------------
        if (!is.null(Ref)) {
            if (!identical(length(Ref), NCOL(Mat))) {
                stop("length(Ref) must be equal to NCOL(Mat)")
            }
        }
        
        if (is.null(colnames(Mat)) &&
            is.null(Wave)) {
            stop("Spectral variable must be specified at Wave or in
                 colnames(Mat)")
        }
        
        if (N <= 0 || N%%1 != 0) {
            stop("N must be a positive, non-zero integer")
        }
        if (corenumber <= 0 || corenumber%%1 != 0) {
            stop("corenumber must be a positive, non-zero integer")
        }
        
        if (!is.logical(preview)) {
            stop("preview needs to logical")
        }
        
        # create an R session for every detected CPU core for parallel computing
        cl <- parallel::makeCluster(corenumber)
        doParallel::registerDoParallel(cl)
        
        # define Wave from colnames if needed ---------------------------------
        if (is.null(Wave)) {
            Wave <- as.numeric(colnames(Mat))
        }
        
        # spectral variable must be increasing, also switch Ref if needed ------
        if (is.unsorted(Wave)) {
            Wave <- Wave[length(Wave):1]
            Mat <- Mat[, NCOL(Mat):1]
            if (!is.null(Ref)) {
                Ref <- Ref[length(Ref):1]
            }
        }
        
        D1 <- dim(Mat)

        # Interpolate values of perturbation variable for equidistant ---------
        # datapoints for fft --------------------------------------------------
        if ((is.null(Time) == FALSE) & (length(Time) == D1[1])) {
            cat(c(format(Sys.time(), "%X -"), "Interpolate Time from",
                  min(Time), "to", max(Time), "\n", "to obtain", N,
                  "equidistant datapoints for FFT", "\n"))
            TIME <- seq(min(Time), max(Time), length.out = N)
            
            tmp <- apply(Mat, 2, function(y) Int(x = Time, y = y))
            Mat <- sapply(tmp, function(x) x(TIME))

            Time <- TIME
        }
        # Get perturbation variables from rownames ----------------------------
        if (is.null(Time) && !is.null(rownames(Mat))) {
            Time <- as.numeric(rownames(Mat))
        }
        
        # Subtract reference -------------------------------------------------
        if (is.null(Ref)) {
            cat(c(format(Sys.time(), "%X -"), "using mean values as reference\n"))
            Ref <- tavspec <- colMeans(Mat)
        } else {
            tavspec <- colMeans(Mat)
        }
        Mat <- sweep(Mat, 2, Ref, "-")
        
        # Apply scaling -------------------------------------------------------
        if (scaling > 0) {
            cat(c(format(Sys.time(), "%X -"), "apply scaling\n"))
            sd1 <- apply(Mat, 1, sd)
            Mat <- Mat / (sd1^scaling)
        }
        
        Syncorr <- matrix(Norm * parallel::parCapply(cl, Mat, get("%*%"), Mat),
                           NCOL(Mat), NCOL(Mat), byrow = T)

        cat(c(format(Sys.time(), "%X -"), "Done calculating synchronous 2D spectrum\n"))
        
        parallel::stopCluster(cl)
        
        # build the sum weigthed sum spectrum
        su <- colSums(1:NROW(Mat) * Mat)
        
        # calculate characteristic perturbation index k and -------------------
        # characteristic perturbation of the distribution t -------------------
        k <- 1 / (NROW(Mat) * tavspec) * su  + (NROW(Mat) + 1) / 2
        dist <- (max(Time) - min(Time)) * (k - 1) / (NROW(Mat) - 1) + min(Time)
        
        # calculate total joint variance tjv from synchronous correlation spectrum
        tjv <- sqrt(outer(diag(Syncorr), diag(Syncorr)))
        tmat <- -1 * outer(dist, dist, "-")
        
        # calculate asynchronous and synchronous codistribution spectra -------
        # from total joint variance tjv and characteristic perturbation of the 
        # distribution t, respectivly tmat ------------------------------------
        Asyncds <- (tmat * tjv) / (max(Time) - min(Time))
        Syncds <- sqrt(tjv^2 - Asyncds^2)
        FT <- Syncds + Asyncds * 1i
        cat(c(format(Sys.time(), "%X -"), "Finished\n"))
        
        Obj <- list(FT = FT, corr = Syncorr, Ref1 = Ref, Ref2 = Ref,
                    Wave1 = Wave, Wave2 = Wave, Time = Time)
        
        # 3d preview of the asynchronous codistribution spectrum --------------
        if (preview == TRUE) {
            if (dim(Asyncds)[1] > 700) {
                tmp1 <- round(seq(1, dim(Asyncds)[1], length = 700))
            } else {
                tmp1 <- seq(1, dim(Asyncds)[1], 1)
            }
            rgl::persp3d(Wave[tmp1], Wave[tmp1], Asyncds[tmp1, tmp1],
                          col = "grey")
        }
        
        class(Obj) <- "corr2d"
        return(Obj)
    }

R/corr2t2d.R
#' Two-trace two-dimensional (2T2D) correlation spectroscopy
#' 
#' \code{corr2t2d} compares a pair of spectra in the form of a cross
#'      correlation analysis.
#'
#' \code{corr2t2d} implements the Two-trace two-dimensional (2T2D) approach
#'     as described by I. Noda (2018) <DOI:10.1016/j.molstruc.2018.01.091>.
#'     The idea is to compare two spectra in a 2D correlation-like
#'     approach which was previously not possible as 2D correlation analysis
#'     usually needs at least three spectra.
#'     
#' @param Sam Numeric vector containing the sample spectrum to be correlated.
#'     Can contain the spectral variable of the sample and reference spectrum
#'     as \code{names}.
#' @param Ref Numeric vector containing the sample spectrum to be correlated.
#'     Can contain the spectral variable of the sample and reference spectrum
#'     as \code{names}.
#' @param Wave Numeric vector containing the spectral variable. Needs to be
#'     specified if names of \code{Sam} and \code{Ref} are undefined.
#' @param preview Logical: Should a 3D preview of the asynchronous codistribution
#'     spectrum be drawn at the end? Uses \code{\link[rgl]{persp3d}} from \pkg{rgl}
#'     package.
#'     
#' @return \code{corr2t2d} returns a list of class "corr2d" containing the
#'     complex correlation matrix (\code{$FT}), the correlation and
#'     disrelation coefficient as a complex matrix ($coef), the sample
#'     \code{$Ref1} and reference spectrum \code{$Ref2} as well as the 
#'     spectral variable \code{$Wave1} and \code{$Wave2}.
#' 
#' @references
#'       I. Noda (2018) <DOI:10.1016/j.molstruc.2018.01.091>
#'       
#' @seealso For plotting of the resulting list containing the 2D correlation
#'     spectra or correlation coefficient see \code{\link{plot_corr2d}} and
#'     \code{\link{plot_corr2din3d}}.
#' 
#' @examples
#'     testdata <- sim2ddata()
#'     
#'     twodtest <- corr2t2d(testdata[4, ], testdata[5, ])
#'     
#'     plot_corr2d(twodtest, Im(twodtest$FT))
#' 
#' @export
corr2t2d <- function(Sam, Ref, Wave = NULL, preview = FALSE)
    {
    if (!identical(length(Sam), length(Ref))) {
      stop("length(Sam) and length(Ref) must be equal")
    }

    if (is.null(names(Sam)) && is.null(names(Ref)) &&
        is.null(Wave)) {
      stop("Spectral variable must be specified at Wave,
           names(Sam) or names(Ref)")
    }
  
    if(is.null(Wave) && is.null(names(Sam))) {
      Wave1 <- as.numeric(names(Ref))
      Wave2 <- as.numeric(names(Ref))
    } else if(is.null(Wave)) {
      Wave1 <- as.numeric(names(Sam))
      Wave2 <- as.numeric(names(Sam))
    } else {
      Wave1 <- Wave
      Wave2 <- Wave
    }
  
    # Calculate 2T2D spectra
    syn2t2d <- Sam %o% Sam + Ref %o% Ref
    asyn2t2d <- Sam %o% Ref - Ref %o% Sam
    
    # Caluclate 2T2D correlation and disrelation coefficient
    corrcoef <- syn2t2d / sqrt(diag(syn2t2d) %o% diag(syn2t2d))
    disrcoef <- asyn2t2d / sqrt(diag(syn2t2d) %o% diag(syn2t2d))
    
    Obj <- list(FT = syn2t2d + asyn2t2d*1i, coef = corrcoef + disrcoef*1i,
                Ref1 = Sam, Ref2 = Ref, Wave1 = Wave1, Wave2 = Wave2)
  
    # 3d preview of the asynchronous 2T2D spectrum ----------------------------
    if (preview == TRUE) {
      if (dim(asyn2t2d)[1] > 700) {
        tmp1 <- round(seq(1, dim(asyn2t2d)[1], length = 700))
      } else {
        tmp1 <- seq(1, dim(asyn2t2d)[1], 1)
      }
      if (dim(asyn2t2d)[2] > 700) {
        tmp2 <- round(seq(1, dim(asyn2t2d)[2], length = 700))
      } else {
        tmp2 <- seq(1, dim(asyn2t2d)[2], 1)
      }
      rgl::persp3d(Wave1[tmp1], Wave2[tmp2], asyn2t2d[tmp1, tmp2], col = "grey")
    }
  
    class(Obj) <- "corr2d"
    return(Obj)
}


R/corr2d.R
#' Two-dimensional correlation analysis.
#'
#' \code{corr2d} calculates the synchronous and asynchronous correlation
#'     spectra between \code{Mat1} and \code{Mat1} (homo correlation)
#'     or between \code{Mat1} and \code{Mat2} (hetero correlation).
#'
#' \code{corr2d} uses a parallel fast Fourier transformation
#'     (\code{\link[stats]{fft}}) to calculate the complex correlation matrix.
#'     For parallelization the \code{\link[foreach]{foreach}} function is used.
#'     Large input matrices (> 4000 columns) can lead to long calculation times
#'     depending on the number of cores used. Also note that the resulting
#'     matrix can become very large, adjust the RAM limit with
#'     \code{\link[utils]{memory.limit}} accordingly. For a detailed description
#'     of the underlying math see references.
#'
#' @param Mat1,Mat2 Numeric matrix containing the data which will be correlated;
#'     '\emph{spectral variable}' by columns and '\emph{perturbation variables}'
#'     by rows. For hetero correlations \code{Mat1} and \code{Mat2} must have
#'     the same number of rows.
#' @param Ref1,Ref2 Numeric vector containing a single spectrum, which will be
#'     subtracted from \code{Mat1} (or \code{Mat2}, respectively) to generate dynamic spectra
#'     for 2D correlation analysis. Default is \code{NULL} in which case the \code{colMeans()}
#'     of \code{Mat1} (or \code{Mat2}, respectively) is used as reference. The length of \code{Ref1}
#'     (or \code{Ref2}) needs to be equal to the number of columns in \code{Mat1} (or \code{Mat2}).
#' @param Wave1,Wave2 Numeric vector containing the spectral variable. Needs to be
#'     specified if column names of \code{Mat1} (or \code{Mat2}) are undefined.
#' @param Time Numeric vector containing the perturbation variables. If specified, \code{Mat1}
#'     (and \code{Mat2} if given) will be interpolated to \code{N} equally spaced perturbation
#'     varibales using \code{Int} to speed up the fft algorithm.
#' @param Int Function specifying how the dataset will be interpolated to give
#'     \code{N} equally spaced perturbation variables. \code{\link[stats]{splinefun}}
#'     (default) or \code{\link[stats]{approxfun}} can for example be used.
#' @param N Positive, non-zero integer specifying how many equally spaced
#'     perturbation variables should be interpolated using \code{Int}. \code{N}
#'     should be higher than 4.\code{corr2d} is fastest if \code{N} is a power
#'     of 2.
#' @param Norm A number specifying how the correlation matrix should be
#'     normalized.
#' @param scaling Positive real number used as exponent when scaling the dataset
#'     with its standard deviation. Defaults to 0 meaning no scaling. 0.5
#'     (\emph{Pareto scaling}) and 1 (\emph{Pearson scaling}) are commonly used to enhance
#'     weak correlations relative to strong correlations.
#' @param corenumber Positive, non-zero integer specifying how many CPU cores
#'     should be used for parallel fft computation.
#' @param preview Logical: Should a 3D preview of the synchronous correlation
#'     spectrum be drawn at the end? Uses \code{\link[rgl]{persp3d}} from \pkg{rgl}
#'     package.
#'     
#' @return \code{corr2D} returns a list of class "corr2d" containing the complex
#'     correlation matrix (\code{$FT}), the used reference spectra (\code{$Ref1},
#'     \code{$Ref2}), the spectral variables (\code{$Wave1}, \code{$Wave2}), the
#'     Fourier transformed data (\code{$ft1}, \code{$ft2}), the (interpolated)
#'     perturbation variables (\code{$Time}) and logical variable (\code{$Het})
#'     indicating if homo (\code{FALSE}) or hetero (\code{TRUE}) correlation was done.
#' 
#' @references 
#'     I. Noda (1993) <DOI:10.1366/0003702934067694>\cr
#'     I. Noda (2012) <DOI:10.1016/j.vibspec.2012.01.006>\cr
#'     R. Geitner et al. (2019) <DOI:10.18637/jss.v090.i03>
#'     
#' 
#' @seealso For plotting of the resulting list containing the 2D correlation
#'     spectra see \code{\link{plot_corr2d}} and \code{\link{plot_corr2din3d}}.
#' 
#' @examples
#'     data(FuranMale, package = "corr2D")
#'     twod <- corr2d(FuranMale, Ref1 = FuranMale[1, ], corenumber = 1)
#'     
#'     plot_corr2d(twod, xlab = expression(paste("relative Wavenumber" / cm^-1)),
#'                       ylab = expression(paste("relative Wavenumber" / cm^-1)))
#'
#' @aliases corr2d
#' 
#' @export
#' @importFrom foreach %dopar%
#' @importFrom stats fft sd
corr2d <-
    function(Mat1, Mat2 = NULL, Ref1 = NULL, Ref2 = NULL,
             Wave1 = NULL, Wave2 = NULL, Time = NULL, Int = stats::splinefun,
             N = 2^ceiling(log2(NROW(Mat1))), Norm = 1/(pi * (NROW(Mat1) - 1)), 
             scaling = 0, corenumber = parallel::detectCores(), preview = FALSE)
    {
        # check user input for errors -----------------------------------------
        if (!is.null(Ref1)) {
            if (!identical(length(Ref1), NCOL(Mat1))) {
                stop("length(Ref1) must be equal to NCOL(Mat1)")
            }
        }
        if (!is.null(Mat2) && !is.null(Ref2)) {
            if (!identical(length(Ref2), NCOL(Mat2))) {
                stop("length(Ref2) must be equal to NCOL(Mat2)")
            }
        }
        
        if (is.null(colnames(Mat1)) &&
            is.null(Wave1)) {
            stop("Spectral variable 1 must be specified at Wave1 or in
                 colnames(Mat1)")
        }
        if (!is.null(Mat2)) {
            if (is.null(colnames(Mat2)) &&
                is.null(Wave2)) {
                stop("Spectral variable 2 must be specified at Wave2 or in
                     colnames(Mat2)")
            }
        }
        
        if (N <= 0 || N%%1 != 0) {
            stop("N must be a positive, non-zero integer")
        }
        if (corenumber <= 0 || corenumber%%1 != 0) {
            stop("corenumber must be a positive, non-zero integer")
        }
        
        if (!is.logical(preview)) {
            stop("preview needs to be logical")
        }
        
        # create an R session for every detected CPU core for parallel computing
        cl <- parallel::makeCluster(corenumber)
        doParallel::registerDoParallel(cl)
        
        # define Wave1/2 from colnames if needed ------------------------------
        if (is.null(Wave1)) {
            Wave1 <- as.numeric(colnames(Mat1))
        }
        if (!is.null(Mat2) && !identical(Mat1, Mat2) && is.null(Wave2)) {
            Wave2 <- as.numeric(colnames(Mat2))
        }
        
        # spectral variable must be increasing, also switch Ref if needed ------
        if (is.unsorted(Wave1)) {
            Wave1 <- Wave1[length(Wave1):1]
            Mat1 <- Mat1[, NCOL(Mat1):1]
            if (!is.null(Ref1)) {
                Ref1 <- Ref1[length(Ref1):1]
            }
        }
        if (!is.null(Mat2) && !identical(Mat1, Mat2) && is.unsorted(Wave2)) {
            Wave2 <- Wave2[length(Wave2):1]
            Mat2 <- Mat2[, NCOL(Mat2):1]
            if (!is.null(Ref2)) {
                Ref2 <- Ref2[length(Ref2):1]
            }
        }
        
        # define Wave2 (if needed), D1/2, Het ---------------------------------
        if (is.null(Mat2) || identical(Mat1, Mat2)) {
            cat(c("HOMO-Correlation:", corenumber, "cores used for calculation\n"))
            Het <- FALSE
            D1 <- D2 <- dim(Mat1)
            Wave2 <- Wave1
            Mat2 <- Mat1
            Ref2 <- Ref1
        } else {
            cat(c("HETERO-Correlation:", corenumber, "cores used for calculation\n"))
            Het <- TRUE
            D1 <- dim(Mat1)
            D2 <- dim(Mat2)
        }
        
        # Interpolate values of perturbation variable for equidistant ---------
        # datapoints for fft --------------------------------------------------
        if ((is.null(Time) == FALSE) & (length(Time) == D1[1])) {
            cat(c(format(Sys.time(), "%X -"), "Interpolate Time from",
                  min(Time), "to", max(Time), "\n", "to obtain", N,
                  "equidistant data points for FFT", "\n"))
            TIME <- seq(min(Time), max(Time), length.out = N)
            
            tmp <- apply(Mat1, 2, function(y) Int(x = Time, y = y))
            Mat1 <- sapply(tmp, function(x) x(TIME))
            
            if (Het == FALSE) {
                Mat2 <- Mat1
            } else {
                tmp <- apply(Mat2, 2, function(y) Int(x = Time, y = y))
                Mat2 <- sapply(tmp, function(x) x(TIME))
            }
            Time <- TIME
        }
        # Get perturbation variables from rownames ----------------------------
        if (is.null(Time) && !is.null(rownames(Mat1))) {
            Time <- as.numeric(rownames(Mat1))
        }
        
        # Subtract reference -------------------------------------------------
        if (is.null(Ref1)) {
            cat(c(format(Sys.time(), "%X -"), "using mean values as reference\n"))
            Ref1 <- colMeans(Mat1)
        }
        Mat1 <- sweep(Mat1, 2, Ref1, "-")
        if (Het == FALSE) {
            Mat2 <- Mat1
            Ref2 <- Ref1
        } else {
          if (is.null(Ref2)) {
            Ref2 <- colMeans(Mat2)
          }
            Mat2 <- sweep(Mat2, 2, Ref2, "-")
        }
        
        # Apply scaling -------------------------------------------------------
        if (scaling > 0) {
            cat(c(format(Sys.time(), "%X -"), "apply scaling\n"))
            sd1 <- apply(Mat1, 1, sd)
            sd2 <- apply(Mat2, 1, sd)
            Mat1 <- Mat1 / (sd1^scaling)
            Mat2 <- Mat2 / (sd2^scaling)
        }
        
        # Do fft for every wavenumber in both matrices to obtain frequency ----
        # dependent vectors for every wavenumber and scalar-multiply them -----
        # for every point in the 2D-correlation spectrum ----------------------
        i <- NULL # Dummy to trick R CMD check 
        FT <- NULL
        cat(c(format(Sys.time(), "%X -"), "Fast Fourier Transformation and multiplication \n",
              "to obtain a", D1[2], "x", D2[2], "correlation matrix", "\n"))
        
        FT <- matrix(NA, NCOL(Mat1), NCOL(Mat2))
        # Selection of fft elements
        ftele1 <- if(NROW(Mat1) %% 2 == 0) {
            1:(NROW(Mat1) - 1) %/% 2 + 1
          } else {
            1:(NROW(Mat1)) %/% 2 + 1
          }
        
        ft1 <- foreach::foreach(i = 1:NCOL(Mat1), .combine = 'cbind') %dopar% {
            fft(Mat1[, i])[ftele1]
        }
        if (Het == FALSE) {
            ft2 <- ft1
        } else {
          # Selection of fft elements
          ftele2 <- if(NROW(Mat2) %% 2 == 0) {
            1:(NROW(Mat2) - 1) %/% 2 + 1
          } else {
            1:(NROW(Mat2)) %/% 2 + 1
          }
          
          ft2 <- foreach::foreach(i = 1:NCOL(Mat2), .combine = 'cbind') %dopar% {
              fft(Mat2[, i])[ftele2]
          }
        }
        
        FT <- matrix(Norm * parallel::parCapply(cl, ft1, get("%*%"), Conj(ft2)), NCOL(ft1), NCOL(ft2), byrow = TRUE)
        cat(c(format(Sys.time(), "%X -"), "Done\n"))
        
        Obj <- list(FT = FT, Ref1 = Ref1, Ref2 = Ref2,
                  Wave1 = Wave1, Wave2 = Wave2, ft1 = ft1, ft2 = ft2,
                  Time = Time, Het = Het)
        
        parallel::stopCluster(cl)
        
        # 3d preview of the synchronous correlation spectrum ------------------
        if (preview == TRUE) {
            if (dim(FT)[1] > 700) {
                tmp1 <- round(seq(1, dim(FT)[1], length = 700))
            } else {
                tmp1 <- seq(1, dim(FT)[1], 1)
            }
            if (dim(FT)[2] > 700) {
                tmp2 <- round(seq(1, dim(FT)[2], length = 700))
            } else {
                tmp2 <- seq(1, dim(FT)[2], 1)
            }
            rgl::persp3d(Wave1[tmp1], Wave2[tmp2], Re(FT)[tmp1, tmp2], col = "grey")
        }
        
        class(Obj) <- "corr2d"
        return(Obj)
    }

#' @method summary corr2d
#' @export
summary.corr2d <- function(object, ...)
{
    cat(c(NROW(object$FT), "x", NCOL(object$FT), if (object$Het == TRUE){"Hetero"} else {"Homo"}, "correlation spectra\n"))
    if (object$Het == TRUE) {
        cat(c("Spectral variable 1:", min(object$Wave1), "-", max(object$Wave1), "\n"))
        cat(c("Spectral variable 2:", min(object$Wave2), "-", max(object$Wave2), "\n"))
    } else {
        cat(c("Spectral variable:", min(object$Wave1), "-", max(object$Wave1), "\n"))
    }
    if (!is.null(object$Time)) {
        cat(c("Perturbation variable:", length(object$Time), "values,", min(object$Time), "-", max(object$Time), "\n"))
    }
    
}

#' @method print corr2d
#' @export
print.corr2d <- function(x, ...)
{
    print.default(x, ...)
  
}

#' Check for object class "corr2d"
#'
#' The function checks if an object is of class "corr2d".
#' 
#' The function uses the \code{\link{inherits}} function.
#' 
#' @param x An object which should be check if it is of class "corr2d".
#' 
#' @return A logical scalar
#'
#' @examples 
#'     data(FuranMale, package = "corr2D")
#'     twod <- corr2d(FuranMale, Ref1 = FuranMale[1, ], corenumber = 1)
#'     
#'     # TRUE
#'     is.corr2d(twod) 
#'     # FALSE
#'     is.corr2d(2) 
#'
#' @references
#'     R. Geitner et al. (2019) <DOI:10.18637/jss.v090.i03>
#'
#' @export
is.corr2d <- function(x)
{
    inherits(x, "corr2d", which = FALSE)
}

